packages/http/src/shared.ts

```diff
@@ -5,31 +5,187 @@ export type TActivityResponse = definitions["v1ActivityResponse"];
   }) {
     const { message, cause, activityId, activityStatus, activityType } = input;
     super(message);
 
     this.name = "TurnkeyActivityError";
-    this.activityId = activityId ?? null;
-    this.activityStatus = activityStatus ?? null;
-    this.activityType = activityType ?? null;
-    this.cause = cause ?? null;
+    this.activityId = activityId ?? undefined;
+    this.activityStatus = activityStatus ?? undefined;
+    this.activityType = activityType ?? undefined;
+    this.cause = cause ?? undefined;
   }
 }
 
+export class TurnkeyActivityConsensusNeededError extends Error {
+  activityId: TActivityId | undefined;
+  activityStatus: TActivityStatus | undefined;
+  activityType: TActivityType | undefined;
+  cause: Error | undefined;
+
+  constructor(input: {
+    message: string;
+    cause?: Error | undefined;
+    activityId?: TActivityId | undefined;
+    activityStatus?: TActivityStatus | undefined;
+    activityType?: TActivityType | undefined;
+  }) {
+    const { message, cause, activityId, activityStatus, activityType } = input;
+    super(message);
+
+    this.name = "TurnkeyActivityConsensusNeededError";
+    this.activityId = activityId ?? undefined;
+    this.activityStatus = activityStatus ?? undefined;
+    this.activityType = activityType ?? undefined;
+    this.cause = cause ?? undefined;
+  }
+}
+
+export function checkActivityStatus(input: {
+  id: string;
+  status: TActivityStatus;
+}) {
+  const { id: activityId, status: activityStatus } = input;
+
+  if (activityStatus === "ACTIVITY_STATUS_CONSENSUS_NEEDED") {
+    throw new TurnkeyActivityConsensusNeededError({
+      message: "Activity requires consensus",
+      activityId,
+      activityStatus,
+    });
+  }
+
+  if (activityStatus !== "ACTIVITY_STATUS_COMPLETED") {
+    throw new TurnkeyActivityError({
+      message: `Expected COMPLETED status, got ${activityStatus}`,
+      activityId,
+      activityStatus,
+    });
+  }
+
+  return true;
+}
+
 export function stableStringify(input: Record<string, any>): string {
   return JSON.stringify(input);
 }
+
+export function assertNonNull<T>(input: T | null | undefined): T {
+  if (input == null) {
+    throw new Error(`Got unexpected ${JSON.stringify(input)}`);
+  }
+
+  return input;
+}
+
+export const TERMINAL_ACTIVITY_STATUSES: definitions["v1ActivityStatus"][] = [
+  "ACTIVITY_STATUS_COMPLETED",
+  "ACTIVITY_STATUS_FAILED",
+  "ACTIVITY_STATUS_REJECTED",
+];
+
+/**
+ * This function is a helper method to easily extract a signature string from a completed signing activity.
+ * Particularly useful for scenarios where a signature requires consensus
+ *
+ * @param activity the signing activity
+ * @return signature {r, s, v}
+ */
+export function getSignatureFromActivity(activity: TActivity): TSignature {
+  if (
+    ![
+      "ACTIVITY_TYPE_SIGN_RAW_PAYLOAD",
+      "ACTIVITY_TYPE_SIGN_RAW_PAYLOAD_V2",
+    ].includes(activity.type)
+  ) {
+    throw new TurnkeyActivityError({
```

Comment:
I'm also thinking this should not be called "TurnkeyActivityError". It's likely an SDK-level / user-level issue if this function is called on the wrong activity type. We might need a plain `BadArgumentError`? Otherwise users will be led to believe the issue is within the turnkey activity processing when they get this error.